### PR TITLE
Oakville branch

### PR DESCRIPTION
**Pull request by zaidqureshi2**

### CODE REVIEW

File: investment-api/src/main/java/com/price/investment/api/models/AccountActivity.java [hunk 1]
```diff
@@ -47,6 +49,13 @@
 	@JoinColumn(name="userId")
 	private User user;
 	
+	@ManyToOne
+    @JoinColumn(name="investmentId")
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/AccountActivity.java]: one activity will have one linked investment

Comment (zaidqureshi2) [reply investment-api/src/main/java/com/price/investment/api/models/AccountActivity.java]: also format code

File: investment-api/src/main/java/com/price/investment/api/models/AuditModel.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package com.price.investment.api.models;
+
+import java.time.LocalDate;
+
+import lombok.Data;
+
+
+@Data
+
+public abstract class AuditModel {
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/AuditModel.java]: all models need to extend this class: see AuditModel in this article https://www.callicoder.com/spring-boot-jpa-hibernate-postgresql-restful-crud-api-example/

File: investment-api/src/main/java/com/price/investment/api/models/Investment.java [hunk 3]
```diff
@@ -41,5 +44,9 @@
 	@NotBlank
 	private LocalDate returnDate;
 
-
+	 @OneToMany(mappedBy = "investmentId")
+	 private List<Project> project = new ArrayList<>();
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/Investment.java]: an investment can only belong to one project, a project can have many projects, its a ManyToOne relationship, mapped by projectId

File: investment-api/src/main/java/com/price/investment/api/models/PaymentMethod.java [hunk 4]
```diff
@@ -68,6 +67,10 @@
 	@OneToOne(cascade = {CascadeType.ALL})
 	@JoinColumn(name="addressId")
 	private Address address;
+	
+	@OneToOne(cascade = {CascadeType.ALL})
+	@JoinColumn(name="userAccountId")
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/PaymentMethod.java]: you don't need to reference the userAccount from here

File: investment-api/src/main/java/com/price/investment/api/models/User.java [hunk 5]
```diff
@@ -46,6 +47,12 @@
 	@OneToOne(cascade = {CascadeType.ALL})
 	@JoinColumn(name="addressId")
 	private Address address;
-
 	
+	@ManyToOne
+    @JoinColumn(name="investmentId")
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/User.java]: one to many relationship, a user can have many investments

File: investment-api/src/main/java/com/price/investment/api/models/User.java [hunk 6]
```diff
@@ -46,6 +47,12 @@
...
 	@JoinColumn(name="addressId")
 	private Address address;
-
 	
+	@ManyToOne
+    @JoinColumn(name="investmentId")
+	private Investment investment;
+
+	@ManyToOne
+    @JoinColumn(name="userAccountId")
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/User.java]: one to one relationship joined by the userId

File: investment-api/src/main/java/com/price/investment/api/models/AccountActivity.java (hunk 1 above)
Comment (yudurrani) [reply investment-api/src/main/java/com/price/investment/api/models/AccountActivity.java]: i think i was confused with one to many relation .. i think those filled diamond to simple are just  parent and child but nothing to do with one to many?

.

File: investment-api/src/main/java/com/price/investment/api/models/Investment.java (hunk 3 above)
Comment (yudurrani) [reply investment-api/src/main/java/com/price/investment/api/models/Investment.java]: thats what was my question too ..

File: investment-api/src/main/java/com/price/investment/api/models/AccountActivity.java (hunk 1 above)
Comment (zaidqureshi2) [reply investment-api/src/main/java/com/price/investment/api/models/AccountActivity.java]: yes, parent child, but a parent can have many children of the same type, which is implied project has many investments, when you do a transaction you will make a single investment etc. I have also written some of them down in the diagram:
image

File: investment-api/src/main/java/com/price/investment/api/models/AccountActivity.java [hunk 7]
```diff
@@ -17,39 +19,41 @@
 import lombok.Data;
 
 @Entity
-@Data
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/AccountActivity.java]: did you mean to delete this?

File: investment-api/src/main/java/com/price/investment/api/models/AuditModel.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+@SuppressWarnings("serial")
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+@Data
+
+public abstract class AuditModel implements Serializable {
+	
+    @Column(name = "createdBy", nullable = false, updatable = false)
+	private String createdBy;
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/AuditModel.java]: please format

File: investment-api/src/main/java/com/price/investment/api/models/Investment.java [hunk 9]
```diff
@@ -1,45 +1,52 @@
...
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.OneToMany;
 import javax.persistence.SequenceGenerator;
 import javax.validation.constraints.NotBlank;
 
 import lombok.Data;
 
 @Entity
-@Data
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/Investment.java]: did you mean to remove this?

File: investment-api/src/main/java/com/price/investment/api/models/Investment.java [hunk 10]
```diff
@@ -1,45 +1,52 @@
...
 	@Id
-    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "investment_seq")
-    @SequenceGenerator(name = "investment_seq", sequenceName = "investment_sequence", allocationSize = 1)
+	@GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "investment_seq")
+	@SequenceGenerator(name = "investment_seq", sequenceName = "investment_sequence", allocationSize = 1)
 	private long id;
-	
-	
+
+	@NotBlank
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/Investment.java]: can you make this generate automatically like account activity

File: investment-api/src/main/java/com/price/investment/api/models/User.java [hunk 11]
```diff
@@ -18,34 +23,38 @@
...
-	@OneToOne(cascade = {CascadeType.ALL})
-	@JoinColumn(name="addressId")
+
+	@OneToOne(cascade = { CascadeType.ALL })
+	@JoinColumn(name = "addressId")
 	private Address address;
 
-	
+	@OneToMany(mappedBy = "user")
+	private List<Investment> books = new ArrayList<>();
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/User.java]: books?

File: investment-api/src/main/java/com/price/investment/api/models/User.java [hunk 12]
```diff
@@ -18,34 +23,38 @@
...
+
+	@ManyToOne
+	@JoinColumn(name = "userId")
```
Comment (zaidqureshi2) [inline investment-api/src/main/java/com/price/investment/api/models/User.java]: join user?

**Pull request merged**